File: courseSchedule1.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegrees=new int[numCourses];
        HashMap<Integer,List<Integer>> map=new HashMap<>();
        for(int i=0;i<prerequisites.length;i++)
        {
            indegrees[prerequisites[i][0]]++;
            if(!map.containsKey(prerequisites[i][1]))
            {
                map.put(prerequisites[i][1],new ArrayList<>());
            }
            map.get(prerequisites[i][1]).add(prerequisites[i][0]);
        }
        Queue<Integer> q=new LinkedList<>();
        for(int i=0;i<indegrees.length;i++)
        {
            if(indegrees[i]==0)
                q.add(i);
        }
        while(!q.isEmpty())
        {
            int temp=q.poll();
            List<Integer> list=map.get(temp);
            if(list!=null)
            {
                for(int data: list)
                {
                    indegrees[data]--;
                    if(indegrees[data]==0)
                        q.add(data);
                }
            }
        }
        for(int i=0;i<indegrees.length;i++)
        {
            if(indegrees[i]>0)
                return false;
        }
        return true;
    }
}

// Time Complexity: O(n) , n=numer of couses
// Space Complexity: O(n)

File: course_schedule_1.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        from collections import deque
        """
        :rtype: bool
        :SC:O(V+E) (v=num of vertexes, E=num of edges)
        :TC:O(V+E)
        """
        # base
        if not prerequisites: return True
        
        inDegrees = [0]*numCourses
        graph = {i:set() for i in range(numCourses)}
        
        # computing hashmap graph and indegrees 
        for edge in prerequisites:
            inn=edge[0]
            out=edge[1]
            inDegrees[inn]+=1
            graph[out].add(inn)
        
        q=deque()
        for course_no, in_deg in enumerate(inDegrees):
            if in_deg==0:
                q.append(course_no)
        
        # BFS
        visited_node=0
        while q:
            node=q.popleft()
            visited_node+=1
            for sett_val in graph[node]:
                inDegrees[sett_val]-=1
                if inDegrees[sett_val]==0:
                    q.append(sett_val)
        
        return visited_node == numCourses
    
        
            
            
            
        
        